File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/SelectShapeMouseCommand.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import mono.common.MouseCursor
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Rect
import mono.state.command.CommandEnvironment

/**
 * A [MouseCommand] to select shapes.
 */
internal object SelectShapeMouseCommand : MouseCommand {
    override val mouseCursor: MouseCursor = MouseCursor.DEFAULT

    override fun execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer
    ): MouseCommand.CommandResultType =
        when (mousePointer) {
            is MousePointer.Down -> MouseCommand.CommandResultType.WORKING
            is MousePointer.Drag -> {
                environment.setSelectionBound(
                    Rect.byLTRB(
                        mousePointer.mouseDownPoint.left,
                        mousePointer.mouseDownPoint.top,
                        mousePointer.boardCoordinate.left,
                        mousePointer.boardCoordinate.top
                    )
                )
                MouseCommand.CommandResultType.WORKING
            }

            is MousePointer.Up -> {
                environment.setSelectionBound(null)

                val area = Rect.byLTRB(
                    mousePointer.mouseDownPoint.left,
                    mousePointer.mouseDownPoint.top,
                    mousePointer.boardCoordinate.left,
                    mousePointer.boardCoordinate.top
                )

                val shapes = if (area.width * area.height > 1) {
                    environment.shapeSearcher.getAllShapesInZone(area)
                } else {
                    emptyList()
                }

                if (!mousePointer.isWithShiftKey) {
                    environment.clearSelectedShapes()
                }
                for (shape in shapes) {
                    environment.addSelectedShape(shape)
                }
                MouseCommand.CommandResultType.WORKING
            }

            is MousePointer.Click -> {
                val shapes =
                    environment.shapeSearcher.getShapes(mousePointer.boardCoordinate).toList()
                if (shapes.isNotEmpty()) {
                    val shape = shapes.last()
                    if (mousePointer.isWithShiftKey) {
                        environment.toggleShapeSelection(shape)
                    } else {
                        environment.clearSelectedShapes()
                        environment.addSelectedShape(shape)
                    }
                }
                MouseCommand.CommandResultType.DONE
            }

            is MousePointer.DoubleClick,
            is MousePointer.Move,
            MousePointer.Idle -> MouseCommand.CommandResultType.DONE
        }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/select-shape-mouse-command.ts
import { Rect } from "$libs/graphics-geo/rect";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { type MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A [MouseCommand] to select shapes.
 */
class SelectShapeMouseCommandImpl implements MouseCommand {
    mouseCursor: MouseCursor = MouseCursor.DEFAULT;

    execute(environment: CommandEnvironment, mousePointer: MousePointer): MouseCommand.CommandResultType {
        switch (mousePointer.type) {
            case MousePointerType.DOWN:
                return MouseCommand.CommandResultType.WORKING;
            case MousePointerType.DRAG:
                environment.setSelectionBound(
                    Rect.byLTRB(
                        mousePointer.mouseDownBoardCoordinate.left,
                        mousePointer.mouseDownBoardCoordinate.top,
                        mousePointer.boardCoordinate.left,
                        mousePointer.boardCoordinate.top,
                    ),
                );
                return MouseCommand.CommandResultType.WORKING;
            case MousePointerType.UP: {
                environment.setSelectionBound(null);
                const area = Rect.byLTRB(
                    mousePointer.mouseDownBoardCoordinate.left,
                    mousePointer.mouseDownBoardCoordinate.top,
                    mousePointer.boardCoordinate.left,
                    mousePointer.boardCoordinate.top,
                );
                // 1 cell is empty
                const isEmpty = area.width * area.height === 1;
                // Only select shapes if the area is not empty, otherwise, this is a click event
                const shapes = !isEmpty ? environment.getAllShapesInZone(area) : [];
                if (!mousePointer.isWithShiftKey) {
                    environment.clearSelectedShapes();
                }
                for (const shape of shapes) {
                    environment.addSelectedShape(shape);
                }
                return MouseCommand.CommandResultType.WORKING;
            }
            case MousePointerType.CLICK: {
                const shapes = environment.getShapes(mousePointer.boardCoordinate);
                if (shapes.length > 0) {
                    const shape = shapes[shapes.length - 1];
                    if (mousePointer.isWithShiftKey) {
                        environment.toggleShapeSelection(shape);
                    } else {
                        environment.clearSelectedShapes();
                        environment.addSelectedShape(shape);
                    }
                }
                return MouseCommand.CommandResultType.DONE;
            }

            case MousePointerType.DOUBLE_CLICK:
            case MousePointerType.MOVE:
            case MousePointerType.IDLE:
                return MouseCommand.CommandResultType.DONE;
        }
    }
}

export const SelectShapeMouseCommand = new SelectShapeMouseCommandImpl();
